### PR TITLE
PEPPER-923  Restrict onchistory upload for exited participants

### PR DESCRIPTION
**Pull request by ssettipalli**
PEPPER-923
During the onc history upload, verify if any of the participants are exited/withdrawn and error out.
Error out whole file upload.

### CODE REVIEW

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/elastic/search/ElasticSearch.java [hunk 1]
```diff
@@ -187,12 +187,22 @@ public ElasticSearch getParticipantsWithinRange(String esParticipantsIndex, int
...
+        BoolQueryBuilder boolQuery = getBoolQueryOfParticipantsId(participantIds);
+        return getElasticSearchParticipants(esIndex, participantIds, boolQuery);
+    }
+
+    public ElasticSearch getParticipantsByShortIds(String esIndex, List<String> participantShortIds) {
+        BoolQueryBuilder boolQuery = getBoolQueryOfParticipantsShortId(participantShortIds);
+        return getElasticSearchParticipants(esIndex, participantShortIds, boolQuery);
+    }
+
+    private ElasticSearch getElasticSearchParticipants(String esIndex, List<String> participantIds, BoolQueryBuilder boolQuery) {
```
Comment (ssettipalli) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/elastic/search/ElasticSearch.java:199]: onc history upload is by shortId (hruid) , hence this refactor to reuse existing code by guid lookup

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java [hunk 2]
```diff
@@ -105,9 +107,28 @@ public void upload(String fileContent) {
         validateRows(rows);
         log.info("Validated {} rows for onc history upload", rows.size());
 
+        //error out if any exited participants.
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java]: We already check the upload participants in `OncHistoryUploadService.getParticipantIds`. That is the correct place to check for exited as well, avoiding another iteration and additional code.

I suggest changing `ESParticipantIdProvider.getParticipantIdForShortId` to instead call `ElasticSearchService.getParticipantDocumentByShortId`, do the participant ID validation as it does now, and return a struct/record/class that includes the participant ID and `ElasticSearchParticipantDto`, or whatever part of `ElasticSearchParticipantDto` you need.

Does that make sense?

**Review by denniscunningham — CHANGES_REQUESTED** [review]

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java (hunk 2 above)
Comment (ssettipalli) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java]: OncHistoryUploadService.getParticipantIds does some additional stuff like creating a MR if needed. 
Felt like code might look cluttered even though we save a trip to ES. 
On a second thought since the method is throwing error anyway and does save adding/refactoring additional code. I modified the code to check for EXITED status and throw error. Added some documentation mentioning that we are checking for EXITED participants.
Thank you.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java [hunk 3]
```diff
@@ -125,17 +130,33 @@ public void upload(String fileContent) {
...
+            if (ptpData.isEmpty()) {
+                throw new OncHistoryValidationException("Invalid short ID " + rec.getParticipantTextId());
+            }
+            if (ptpData.get().getStatus().isPresent() && ptpData.get().getStatus().get().startsWith("EXITED")) {
+                exitedParticipants.add(rec.getParticipantTextId());
+                log.info("Participant {} is exited, skipping", rec.getParticipantTextId()); //todo
+            }
+
+            if (!exitedParticipants.isEmpty()) {
+                continue; //skip MR verify/creation. continuing to collect any other exited participant hruids.
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java]: Comments at the end of a line of code do not meet the DSM style guide.
Are you getting a checkstyle warning?

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java [hunk 4]
```diff
@@ -125,17 +130,33 @@ public void upload(String fileContent) {
...
                                                       ParticipantIdProvider participantIdProvider, boolean updateElastic) {
         Map<Integer, Integer> medIds = new HashMap<>();
+        List<String> exitedParticipants = new ArrayList<>();
 
         ParticipantDao participantDao = ParticipantDao.of();
 
+        String esIndex = new DDPInstanceDao().getDDPInstanceByInstanceName(realm).orElseThrow().getEsParticipantIndex();
         for (OncHistoryRecord rec : oncHistoryRecords) {
             int participantId = participantIdProvider.getParticipantIdForShortId(rec.getParticipantTextId());
+            Optional<ElasticSearchParticipantDto> ptpData = elasticSearchService.getParticipantDocumentByShortId(
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java]: This still makes two ES calls, which is unncessary.
Let me know if you want me to provide a commit for this code if my suggestions are unclear.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java [hunk 5]
```diff
@@ -125,17 +130,33 @@ public void upload(String fileContent) {
...
+                    rec.getParticipantTextId(), esIndex);
+            if (ptpData.isEmpty()) {
+                throw new OncHistoryValidationException("Invalid short ID " + rec.getParticipantTextId());
+            }
+            if (ptpData.get().getStatus().isPresent() && ptpData.get().getStatus().get().startsWith("EXITED")) {
+                exitedParticipants.add(rec.getParticipantTextId());
+                log.info("Participant {} is exited, skipping", rec.getParticipantTextId()); //todo
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java]: What is the todo?

**Review by denniscunningham — COMMENTED** [review]
More comments.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java (hunk 5 above)
Comment (ssettipalli) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java]: meant to delete it

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java (hunk 4 above)
Comment (ssettipalli) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java]: done. removed ES call

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java [hunk 6]
```diff
@@ -125,23 +129,38 @@ public void upload(String fileContent) {
...
+            }
+
+            if (!exitedParticipants.isEmpty()) {
+                //skip MR verify/creation. continuing to collect any other exited participant hruids.
+                continue;
+            }
 
+            int participantId;
             try {
+                participantId = ptpData.getDsm().get().getParticipant().get().getParticipantId().intValue();
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java]: Combining all this in one line removes error detail provided in `ESParticipantIdProvider`.
Perhaps break this out into a separate method that replicates the code in `ESParticipantIdProvider.getParticipantIdForShortId`.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java [hunk 7]
```diff
@@ -125,23 +129,38 @@ public void upload(String fileContent) {
...
     protected Map<Integer, Integer> getParticipantIds(List<OncHistoryRecord> oncHistoryRecords,
                                                       ParticipantIdProvider participantIdProvider, boolean updateElastic) {
         Map<Integer, Integer> medIds = new HashMap<>();
+        List<String> exitedParticipants = new ArrayList<>();
 
         ParticipantDao participantDao = ParticipantDao.of();
 
+        String esIndex = new DDPInstanceDao().getDDPInstanceByInstanceName(realm).orElseThrow().getEsParticipantIndex();
         for (OncHistoryRecord rec : oncHistoryRecords) {
-            int participantId = participantIdProvider.getParticipantIdForShortId(rec.getParticipantTextId());
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java:138]: I think there are tests that use `ParticipantIdProvider`. How are they affected with this removal?

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java [hunk 8]
```diff
@@ -152,6 +171,14 @@ protected Map<Integer, Integer> getParticipantIds(List<OncHistoryRecord> oncHist
                     this.realm, updateElastic);
             medIds.put(participantId, medId);
         }
+
+        if (!exitedParticipants.isEmpty()) {
+            log.error("Found {} exited participants: {} in Onc History Upload", exitedParticipants.size(), exitedParticipants);
+            throw new OncHistoryValidationException("One or more of the uploaded onc histories is associated with a withdrawn participant. "
+                    + " Please remove onc histories for these withdrawn participants " + exitedParticipants
+                    + " from the file and upload it again.");
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java]: This will read better if `exitedParticipants` is added at the end of the error string, after a colon.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java (hunk 7 above)
Comment (ssettipalli) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java:138]: hmm . OncHistoryUploadServiceTest.java is setting and has ParticipantIdProvider code . Looks like ParticipantIdProvider was introduced for test support rather than using ElasticSEarchService directly.
Looks like all that need to be refactored now for the test to even run.  Not sure its all worth the time.

Comment (denniscunningham) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java:138]: Right.
If you look at my original suggestion it was to retain `ParticipantIdProvider` with a new `getParticipantIdForShortId` interface.

Are you suggesting disabling tests in  OncHistoryUploadServiceTest?

Comment (ssettipalli) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java:138]: refactored and added test case with EXITED participant

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/ESParticipantIdProvider.java [hunk 9]
```diff
@@ -42,4 +43,14 @@ public int getParticipantIdForShortId(String shortId) {
             throw new DsmInternalError("Invalid dsm.participant.participantId for shortId " + shortId);
         }
     }
+
+    /**
+     * Given a participant short ID return a participant ID
+     * @throws DsmInternalError for bad ES behavior
+     * @throws DSMBadRequestException when no participant ID is found for short ID
+     */
+    public Optional<ElasticSearchParticipantDto> getParticipantDataForShortId(String shortId) {
```
Comment (ssettipalli) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/ESParticipantIdProvider.java:52]: Considered adding method to verifyExitedPtpAndGetId , but since I wanted to collect ALL exited participants shortIds to display back to user, looked like I had to throw exception here, catch and check for that exception in OncHistoryUploadService to move forward. 
To avoid all that, I ended up just getting full ptp data here and doing what I need to do in OncHistoryUploadService. 
I can change if needed.

File: pepper-apis/dsm-core/src/test/resources/onchistory/lmsOncHistoryExited.txt [hunk 10]
```diff
@@ -0,0 +1,4 @@
+RECORD_ID	FACILITY_PATH_REVIEW	FACILITY_PX	ACCESSION	DATE_PX	TYPE_PX	LOCATION_PX	HISTOLOGY	TUMOR_SIZE	VIABLE_TUMOR	NECROSIS	TX_EFFECT	BLOCKS_TO_REQUEST	SLIDES_TO_REQUEST	SLIDES_TOTAL	REQUEST_STATUS
+xyz-exit		my place	4	3/15/2005		Office		5cm x 6cm x 1cm	100	50		abc	"a,b"		review
```
Comment (ssettipalli) [inline pepper-apis/dsm-core/src/test/resources/onchistory/lmsOncHistoryExited.txt:2]: used this ptp as EXITED ptp and mocked data accordingly for testing

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java [hunk 11]
```diff
@@ -383,4 +427,6 @@ public Map<String, OncHistoryUploadColumn> initColumnsForStudy() {
     protected Map<String, OncHistoryUploadColumn> getStudyColumns() {
         return studyColumns;
     }
+
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java]: Is this needed?

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/ESParticipantIdProvider.java [hunk 12]
```diff
@@ -37,9 +50,10 @@ public int getParticipantIdForShortId(String shortId) {
             throw new DsmInternalError("ES returned empty dsm.participant.participantId object for shortId " + shortId);
         }
         try {
-            return participantID.intValue();
+            return ptpData;
         } catch (Exception e) {
             throw new DsmInternalError("Invalid dsm.participant.participantId for shortId " + shortId);
         }
     }
+
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/ESParticipantIdProvider.java:58]: Is this change needed?

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/ESParticipantIdProvider.java [hunk 13]
```diff
@@ -37,9 +50,10 @@ public int getParticipantIdForShortId(String shortId) {
             throw new DsmInternalError("ES returned empty dsm.participant.participantId object for shortId " + shortId);
         }
         try {
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/ESParticipantIdProvider.java]: This try is no longer necessary.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/ESParticipantIdProvider.java [hunk 14]
```diff
@@ -27,6 +28,18 @@ public ESParticipantIdProvider(String realm, String participantIndex) {
      * @throws DSMBadRequestException when no participant ID is found for short ID
      */
     public int getParticipantIdForShortId(String shortId) {
+        //getParticipantDataForShortId does all the validations to make sure participant ID exists
+        return getParticipantDataForShortId(shortId).getDsm().get().getParticipant().get().getParticipantId().intValue();
+    }
+
+    /**
+     * Given a participant short ID return a participant ID
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/ESParticipantIdProvider.java]: This JavaDoc is not correct.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/ESParticipantIdProvider.java [hunk 15]
```diff
@@ -27,6 +28,18 @@ public ESParticipantIdProvider(String realm, String participantIndex) {
      * @throws DSMBadRequestException when no participant ID is found for short ID
      */
     public int getParticipantIdForShortId(String shortId) {
+        //getParticipantDataForShortId does all the validations to make sure participant ID exists
+        return getParticipantDataForShortId(shortId).getDsm().get().getParticipant().get().getParticipantId().intValue();
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/ESParticipantIdProvider.java]: However, you are missing the try block around the `intvalue`.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java [hunk 16]
```diff
@@ -125,23 +126,35 @@ public void upload(String fileContent) {
...
+                exitedParticipants.add(rec.getParticipantTextId());
+            }
+
+            if (!exitedParticipants.isEmpty()) {
+                //skip MR verify/creation. continuing to collect any other exited participant hruids.
+                continue;
+            }
 
+            //since participantIdProvider.getParticipantDataForShortId already made sure participant ID exists.. its ok to get it directly
+            int participantId = ptpData.getDsm().get().getParticipant().get().getParticipantId().intValue();
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java]: You are not checking that `intValue` will work in getParticipantDataForShortId.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java [hunk 17]
```diff
@@ -125,23 +126,35 @@ public void upload(String fileContent) {
...
             try {
                 ParticipantDto participant = participantDao.get(participantId).orElseThrow();
                 rec.setDdpParticipantId(participant.getDdpParticipantId().orElseThrow());
             } catch (Exception e) {
-                throw new DsmInternalError("Participant not found for id " + participantId, e);
+                throw new DsmInternalError("Failed to get Participant/Participant Id for " + rec.getParticipantTextId(), e);
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/onchistory/OncHistoryUploadService.java]: Why change this? `participantId` is available.

**Review by denniscunningham — CHANGES_REQUESTED** [review]

**Review by denniscunningham — APPROVED** [review]

**Comment by ssettipalli** [comment]
Thank you Dennis
